<div>
    <p>The “Annotation Browser” panel displays the annotations in detail for each isoform.</p>
    <p>
        <img src="img/report-browser-annots.png" class="img-thumbnail img-responsive img-center" style="width: 80%">
    </p>
    <h5>Num. Functional Residues</h5>
    <ul>
        <li>the peptide position,</li>
        <li>PDB ligand id,</li>
        <li>and a score of all predicted catalytic residues. It is a measure of prediction reliablity. The score
            should be close to six.</li>
    </ul>

    <h5>Tertiray Structure Score</h5>
    <ul>
        <li>the range of amino acids,</li>
        <li>the best PDB template,</li>
        <li>and the percentage identity of the alignment.</li>
    </ul>

    <h5>Conservation score (vertebrates)</h5>
    <ul>
        <li>the nearest homologue specie,</li>
        <li>and the percentage of identity.</li>
    </ul>
    <h5>Domains score</h5>
    <ul>
        <li>the alignment start,</li>
        <li>the alignment end,</li>
        <li>domain name,</li>
        <li>the best e-value. This e-value could come from other transcripts that shares the peptide region.</li>
        <li>And the Pfam-isoform alignments.</li>
    </ul>
    <h5>Num. Transmembrane Helices</h5>
    <ul>
        <li>helix start,</li>
        <li>helix end,</li>
        <li>and flag if the helix is damaged.</li>
    </ul>

    <h5>Signal Peptide Sequences and Mitochondrial Signal Sequences</h5>
    <ul>
        <li>the type of singal,</li>
        <li>start position of signal,</li>
        <li>and end position of signal.</li>
    </ul>
</div>
