<div>
    <h4>Principal Isoform flags</h4>
    <dl>
        <p>
        <dd><p>APPRIS selects a single CDS variant for each gene as the 'PRINCIPAL' isoform based on the range of
        protein features. Principal isoforms are tagged with the numbers 1 to 5, with 1 being the most reliable. The definition of the flags are as follows:</p>
        </dd>
        <ul>
                <li><strong>PRINCIPAL:1</strong>
                    <p>Transcript(s) expected to code for the main functional isoform based solely on the core modules in the APPRIS database. The APPRIS core modules map protein structural and functional information and cross-species conservation to the annotated variants.</p>
                </li>

                <li><strong>PRINCIPAL:2</strong>
                    <p>Where the APPRIS core modules are unable to choose a clear principal variant (approximately 25% of human protein coding genes), the database chooses two or more of the CDS variants as "candidates" to be the principal variant.</p>
                    <p>If one of these candidates has a distinct CCDS identifier it is selected as the
                        principal variant for that gene. A CCDS identifier shows that there is consensus between RefSeq and GENCODE/Ensembl for that variant, guaranteeing that the variant has cDNA support.</p>
                </li>

                <li><strong>PRINCIPAL:3</strong>
                    <p>Where the APPRIS core modules are unable to choose a clear principal variant and there more than one of the variants have distinct CCDS identifiers, APPRIS selects the variant with lowest CCDS identifier as the principal variant. The lower the CCDS identifier, the earlier it was annotated.</p>
                    <p>Consensus CDS annotated earlier are likely to have more cDNA evidence. Consecutive CCDS identifiers are not included in this flag, since they will have been annotated in the same release of CCDS. These are distinguished with the next flag.</p>
                    <p>In addition, there is more than one variant with a distinct (but consecutive) CCDS identifiers, APPRIS choose the variant when all splice junctions are supported by at least one non-suspect mRNA. This information is reported by the method Transcript Support Level (TSL), which is a method to highlight the well-supported and poorly-supported transcript models for users. The method relies on the primary data that can support full-length transcript structure: mRNA and EST alignments supplied by UCSC and Ensembl.</p>
                </li>

                <li><strong>PRINCIPAL:4</strong>
                    <p>Where the APPRIS core modules are unable to choose a clear principal CDS and there is more than one variant with a distinct (but consecutive) CCDS identifiers and all the splice junctions are not well-supported, APPRIS selects the longest CCDS isoform as the principal variant.</p>
                </li>

                <li><strong>PRINCIPAL:5</strong>
                    <p>Where the APPRIS core modules are unable to choose a clear principal variant and none of the candidate variants are annotated by CCDS, APPRIS selects the longest of the candidate isoforms as the principal variant.</p>
                </li>
            </ul>
        <dd><p>For genes in which the APPRIS core modules are unable to choose a clear principal variant (approximately 25% of human protein coding genes) the "candidate" variants not chosen as principal are labeled in the following way:</p>
        </dd>
        <ul>
            <li><strong>ALTERNATIVE:1</strong>
                <p>Candidate transcript(s) models that are conserved in at least three tested non-primate species.</p>
            </li>

            <li><strong>ALTERNATIVE:2</strong>
                <p>Candidate transcript(s) models that appear to be conserved in fewer than three tested non-primate species.</p>
            </li>
        </ul>
        </p>
        <p>
            Non-candidate transcripts are not flagged and are considered as <strong>"MINOR"</strong> transcripts.
        </p>
    </dl>
</div>